src/fast-path.js

```diff
@@ -245,6 +245,21 @@ FPp.needsParens = function(assumeExpressionContext) {
     return true;
   }
 
+  if (
+    (node.type === "BindExpression" ||
+     node.type === "UpdateExpression" ||
+     node.type === "BinaryExpression" ||
+     node.type === "LogicalExpression" ||
+     node.type === "AssignmentExpression" ||
+     node.type === "ConditionalExpression" ||
+     node.type === "SequenceExpression") &&
+    ((parent.type === "ArrowFunctionExpression" && parent.body === node) ||
+    parent.type === "ExpressionStatement") &&
+    startsWithOpenCurlyBrace(node)
+  ) {
+    return true;
+  }
+
   switch (node.type) {
     case "CallExpression":
       if (
```

Comment:
If you go that route, I think you can then delete all of

```js
    case "ObjectExpression":
      if (parent.type === "ArrowFunctionExpression" && name === "body") {
        return true;
      }
      if (parent.type === "TaggedTemplateExpression") {
        return true;
      }
      if (parent.type === "MemberExpression") {
        return name === "object" && parent.object === node;
      }
```

which, per git blame, looks to me to have been a hack trying to fix the bug which this PR fixes properly.